### DISCUSSION TITLE
What's the purpose of the daemon vs running only fwupdtool

### DISCUSSION THREAD

**Discussion by ronny-rentner** [fwupd/fwupd#5187] [General]
It seems you can do everything without having a daemon by just using fwupdtool.

Why would I, as a user, want to have a daemon running all the time, taking resources instead of just having a systemd timer or cronjob to look for new firmware once a day?

**Comment by hughsie** [comment]
> Why would I, as a user, want to have a daemon running all the time

A few reasons:

 * it takes a few tens of seconds to start on typical hardware, and it's not acceptable to wait >10s per request in UI programs
 * The daemon quits on idle when required (or on low memory), and is only autostarted when required
 * Device hotplug needs to work; you don't want to do all the CPU intensive work on each event
 * Some composite devices need to be handled in order; installing "one firmware" is sometimes a dozen devices going away and comping back in various orders and that's not well served with a oneshot command.

**Comment by ronny-rentner** [comment]
Hmmm, regarding your points:

> it takes a few tens of seconds to start on typical hardware, and it's not acceptable to wait >10s per request in UI programs

I don't really see the connection here. For example, I am a Debian user and I use cron-apt to download all updates in the background. When I want to install them, they are cached locally and the wait time for UI programs is zero. Does fwupdtool not use a cache of binary data that it has downloaded?

> The daemon quits on idle when required (or on low memory), and is only autostarted when required

I cannot confirm this is happening on idle, at least not on my Debian Testing desktop computer, but maybe that's a bug in Debian.
```# systemctl status fwupd
● fwupd.service - Firmware update daemon
     Loaded: loaded (/lib/systemd/system/fwupd.service; static)
     Active: active (running) since Sat 2022-10-22 08:32:14 CEST; 20h ago
       Docs: https://fwupd.org/
   Main PID: 3059 (fwupd)
      Tasks: 5 (limit: 38142)
     Memory: 21.9M
        CPU: 5.772s
     CGroup: /system.slice/fwupd.service
             └─3059 /usr/libexec/fwupd/fwupd

Oct 22 08:32:11 yuna systemd[1]: Starting Firmware update daemon...
Oct 22 08:32:14 yuna systemd[1]: Started Firmware update daemon.
```

> Device hotplug needs to work; you don't want to do all the CPU intensive work on each event

I don't really get when you mean with CPU intensive work? Of course CPU intensive work should be cached, also between reboots.

> Some composite devices need to be handled in order; installing "one firmware" is sometimes a dozen devices going away and comping back in various orders and that's not well served with a oneshot command.

Indeed, similar to Debian packages where you have dependencies, etc. It should be easy for any tool, daemon or not, to stick to a predefined order.

**Comment by hughsie** [comment]
¯\_(ツ)_/¯ maybe disable the daemon on your system and use fwupdtool, I don't mind. For the general purpose user a daemon works much better for the usecases we're targeting.
